auraed/src/observe/proc_cache.rs

```diff
@@ -0,0 +1,372 @@
+ *                |  ██║  ██║╚██████╔╝██║  ██║██║  ██║███████╗ |              *
+ *                |  ╚═╝  ╚═╝ ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═╝╚══════╝ |              *
+ *                +--------------------------------------------+              *
+ *                                                                            *
+ *                         Distributed Systems Runtime                        *
+ *                                                                            *
+ * -------------------------------------------------------------------------- *
+ *                                                                            *
+ *   Licensed under the Apache License, Version 2.0 (the "License");          *
+ *   you may not use this file except in compliance with the License.         *
+ *   You may obtain a copy of the License at                                  *
+ *                                                                            *
+ *       http://www.apache.org/licenses/LICENSE-2.0                           *
+ *                                                                            *
+ *   Unless required by applicable law or agreed to in writing, software      *
+ *   distributed under the License is distributed on an "AS IS" BASIS,        *
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ *   See the License for the specific language governing permissions and      *
+ *   limitations under the License.                                           *
+ *                                                                            *
+\* -------------------------------------------------------------------------- */
+
+use crate::ebpf::tracepoint::PerfEventBroadcast;
+use aurae_ebpf_shared::{ForkedProcess, ProcessExit};
+use std::time::SystemTime;
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::Arc,
+    time::Duration,
+};
+use tokio::sync::Mutex;
+
+#[cfg(not(test))]
+pub fn now() -> SystemTime {
+    SystemTime::now()
+}
+#[cfg(test)]
+pub fn now() -> SystemTime {
+    use test_helpers::mock_time;
+
+    mock_time::now()
+}
+
+const PID_MAX: usize = 4194304;
+
+pub trait ProcessInfo {
+    fn get_nspid(&self, pid: i32) -> Option<i32>;
+}
+
+pub(crate) struct ProcfsProcessInfo {}
+
+impl ProcessInfo for ProcfsProcessInfo {
+    fn get_nspid(&self, pid: i32) -> Option<i32> {
+        procfs::process::Process::new(pid)
+            .and_then(|p| p.status())
+            .ok()
+            .and_then(|s| s.nspid)
+            .and_then(|nspid| nspid.last().copied())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct Eviction {
+    pid: i32,
+    evict_at: SystemTime,
+}
+
+/// Cache that allows for accessomg process info (right now only namespace PIDs)
+/// beyond the lifetime of a process.
+///
+/// mention eBPF events
+/// mention eviction strategy
+#[derive(Debug)]
+pub struct ProcCache {
```

Comment:
Instead of sharing a mock `SystemTime` which necessitates using `#[serial]` for the tests, why not have `ProcCache` have a field with the implementation?

Reply:
I thought about this, but because we call this from 2 different background threads, this would be yet another `Arc<Mutex<Clock>>` that we have to clone and wait for, just for testing purposes.

Reply:
I'm not seeing why that would need a that, so I'll assume my comment was not clear. I was trying to say that every instance of `ProcCache` can have it's own instance of `Clock`.

Reply:
Right, but we need to inject that `Clock` from the test and be able to advance it from the tests _and_ read it from those 2 background threads that are polling the channels.